### PR TITLE
Add new information to LRD

### LINKED ISSUE #362: Layout adjustment - PHPdocs and Tags
Hi @kevincobain2000, 
I am making a layout change to meet a need in my project and I would like to know your opinion on submitting this PR for the project or do you think it is not necessary.
My main problem was not having the option to display description and example for the fields of both BODY PARAMETERS and QUERY PARAMETERS.
After displaying these values, the layout became very confusing so I made an adjustment that ended up changing the screen the way it was. The rules are no longer displayed each on a line, but rather all on a single line, and below we have a collapse to display the descriptions and examples. Below is a print of how it turned out. 
Note. I didn't really like the idea of collapse, but it was what I could think of to display the description and example and not clutter the screen, if you have any suggestions I would be very grateful.

before

after

Any suggestion is welcome.

Thank's for your time

### PR DESCRIPTION
**Pull request by fernandopiovezan1**
Add summary to routes
Add description to routes
Add friendly name to route menu
Add description to the field along with the rules
Add example to the field next to the rules
Add tags for grouping routes in OpenAPI export
Changes rules display format from vertical to horizontal (use of space)

resolve #362 

Related issues

https://github.com/rakutentech/laravel-request-docs/issues/301
https://github.com/rakutentech/laravel-request-docs/issues/331

### CODE REVIEW

**Comment by kevincobain2000** [comment]
Thanks, give me some time. I'd need to test it.
Have you run the `phpunit`?

**Comment by fernando-piovezan-instar** [comment]
I couldn't run `phpunit`, the tests fail to get the routes, I'm trying to adjust this but still without success. As soon as I can understand these route problems, I will run `phpunit`
Thanks

**Comment by fernandopiovezan1** [comment]
I had not performed the necessary adjustments to the tests, in this last commit I inserted the new values into the mock, so the tests are passing normally

**Comment by kevincobain2000** [comment]
@fernandopiovezan1 Please rebase from `develop` I have removed the PHP 7.4 from CI matrix https://github.com/rakutentech/laravel-request-docs/commit/fe09ae61180f60459f9a39d558775e1a0c5deec0

**Comment by fernandopiovezan1** [comment]
@kevincobain2000 I did the rebase as requested, now the tests are all ok, I believe it is now possible to do the merge

**Comment by kevincobain2000** [comment]
Thanks. 

Will test and merge this weekend. 
@kitloong

File: README.md [hunk 1]
```diff
@@ -158,6 +158,38 @@ Example of using it in the controller:
...
+# Route grouping
+You can add a `@LRDtags` parameter to the description of your controller to group the routes on the dashboard. Endpoint grouping is still done by routes and these tags will only be used as a friendly name for viewing the dashboard.
+@LRDTags is used to generate documentation in OpenAPI format
+Example of using it in the controller:
+
+```php
+    /**
+     * Class DemoController
+     * @package App\Http\Controllers
+     * @LRDtags Demonstration
```
Comment (kevincobain2000) [inline README.md:170]: This `@LRDtags` if added to one controller, is picked up by all the other controllers too.
As example:

I only added this to the `api/v1/health` controller, while it goes on display to all the other controllers such as `api/v1/holi...`

Screenshot 2024-06-03 at 10 51 28

```
/**
 * Class HealthController
 * Summary of method.
 * @LRDtags Demonstration
 * Demonstration management endpoint
 */
class HealthController extends ApiBaseController
{
    /**
     * Summary of method.
     * Description of method
     * @param HealthIndexRequest $request
     */
    public function index(HealthIndexRequest $request): Resource
```



Actually, I am not in favor of adding `@LRDtags`, the whole purpose of the lib is to not ask users to remember all the syntax sugar such as `@LRDtags` and it could generate the docs based on what a PHP user already knows. 

Summary and description work as intended and are fine!
Considering removal of `@LRDtags`, what do you think? @kitloong and @fernando-piovezan-instar

File: README.md [hunk 2]
```diff
@@ -178,6 +210,69 @@ So, the precedence is `Controller method PHPDoc` < `Rules method PHPDoc` < `Rule
     public function index(MyIndexRequest $request): Resource
     {
 ```
+# Rules and field description and example
+Para obter uma descrição do campos e das regras você pode adicionar um método `fieldDescriptions()` a seu `formRequest` e retornar um array associativo onde a chave é o nome do campo e o valor é a descrição e o exemplo do campo.
```
Comment (kevincobain2000) [inline README.md:214]: English?

**Comment by kevincobain2000** [comment]
I have tested it. The tags automatically being added to all controllers is fixed.

However,
is following the expected behaviour?

I added same `@LRDTags Demonstration` to 2 controllers, HealthController1 and HealthController2.
It does show the correct tag but doesn't group them.

Screenshot 2024-06-04 at 10 37 32

File: src/LaravelRequestDocsToOpenApi.php [hunk 3]
```diff
@@ -111,19 +115,23 @@ protected function attributeIsFile(string $rule): bool
...
         if (is_array($rule)) {
             $rule = implode('|', $rule);
         }
         $parameter = [
-            'name' => $attribute,
-            'description' => $rule,
             'in' => 'query',
+            'name' => $attribute,
+            'description' => $description['description'] ?? $rule,
+            'example' => $description['example'] ?? null,
```
Comment (kitloong) [inline src/LaravelRequestDocsToOpenApi.php:127]: Is there any way for us to validate the OpenAI specification?

File: tests/Stubs/TestControllers/WelcomeController.php [hunk 4]
```diff
@@ -8,6 +8,9 @@
 use Rakutentech\LaravelRequestDocs\Tests\Stubs\TestRequests\WelcomeIndexRequest;
 use Rakutentech\LaravelRequestDocs\Tests\Stubs\TestRequests\WelcomeStoreRequest;
 
+/**
+ * @LRDtags Welcome
```
Comment (kitloong) [inline tests/Stubs/TestControllers/WelcomeController.php:12]: This PR added a lot of features, but the test doesn't cover them at all.

You should add tests and the test should break without your added features

File: ui/src/libs/types.tsx [hunk 5]
```diff
@@ -25,6 +34,9 @@ export interface IAPIInfo {
     group_index: number;
     responses: string[];
     examples: IAPIExample;
+    fieldInfo: IAPIFieldInfos;
```
Comment (kitloong) [inline ui/src/libs/types.tsx:37]:
```suggestion
    field_info: IAPIFieldInfos;
```

File: ui/src/libs/types.tsx [hunk 6]
```diff
@@ -25,6 +34,9 @@ export interface IAPIInfo {
...
+    rulesOrder: string[];
+    tags: string;
```
Comment (kitloong) [inline ui/src/libs/types.tsx:39]: Apply to other naming in this PR too.

If is a list / collection, use plural.
Else use singular.

File: config/request-docs.php [hunk 7]
```diff
@@ -181,4 +181,20 @@
     ],
 
     'use_factory' => false,
+
+    'rules_order' => [
```
Comment (kitloong) [inline config/request-docs.php:185]: May I know why do we need to add this ordering?

If the user want to decided the ordering, can't they define from the `rules` ?

`'page' => 'nullable|integer|min:1',`

**Review by kitloong — CHANGES_REQUESTED** [review]
Hi @fernandopiovezan1, thank you for the PR.

I have done the first round review, but to be honest I stop review at some point due to the amount of changes needed I have seen.

Thank you for your efforts to add the following keys:

1. `summary`
2. `description`
3. `field_info`
4. `rules_order`
5. `tag`
6. `examples`

I like the 1., 2., and 5. implementation, for me they are perfect.

The 3. is okay, but I am being hold back by the idea to add a new method to the Request class, I just feel like it is not the best way to do it.

For 4., I am not sure if we really need it. 

For 6., I feel like it is incomplete, and I wonder how could we make it possible for general use for different use cases. Also, I have some difficulty to picture what is the expected output. 

The test is missing for all the new keys, and I think it is important to have them.
Currently, you are just setting the missing key in the mock json, but it is not enough..

I think I will need to do a second round of review, but I would like to ask you to consider the following:

1. Do we have a better way to implement the `field_info`?
2. Do we really need the `rules_order`?
3. Could you complete the `examples` implementation? Else I would like to suggest to remove it.
4. Cover all features with test, and the test should always failed without your feature.

File: src/LaravelRequestDocs.php [hunk 8]
```diff
@@ -226,15 +247,27 @@ public function getControllersInfo(array $onlyMethods): Collection
...
+        /** @var Doc $doc */
         foreach ($docs as $doc) {
             if ($doc->isClosure()) {
                 // Skip to next if controller is not exists.
                 continue;
             }
 
-            $controllerReflectionMethod = new ReflectionMethod($doc->getControllerFullPath(), $doc->getMethod());
+            try {
+                $controllerReflectionMethod = new ReflectionMethod($doc->getControllerFullPath(), $doc->getMethod());
```
[the remark this answers is not in the document]
Comment (fernandopiovezan1) [reply src/LaravelRequestDocs.php:258]: When using an `apiResource` or a `resource` in the routes and if any method is not found an exception is generated, to prevent this, try catch was used. I'm going to change it so that the try covers all the code within the foreach, this way I believe it will be more comprehensive
image

File: src/LaravelRequestDocsToOpenApi.php (hunk 3 above)
Comment (kevincobain2000) [reply src/LaravelRequestDocsToOpenApi.php:127]: @kitloong is trying to check if the new changes that come to openAPI generated json have been tested using Postman.
I ll test it out, you can skip this comment.

File: src/LaravelRequestDocs.php [hunk 9]
```diff
@@ -325,31 +361,29 @@ public function lrdDocComment(string $docComment): string
...
-                $modelName = Str::replace('APIController', '', $controllerName);
-                $fullModelName = "App\\Models\\" . $modelName;
-
-                if (!class_exists($fullModelName)) {
-                    return;
-                }
+        try {
+            $controllerName = class_basename($doc->getController());
+            $modelName      = Str::replace('APIController', '', $controllerName);
+            $fullModelName  = "App\\Models\\" . $modelName;
```
[the remark this answers is not in the document]
Comment (fernandopiovezan1) [reply src/LaravelRequestDocs.php:367]: In this case, what I could suggest would be two new keys in the configuration, one to inform the path of the application's models (it would be a way for projects that use a different directory for the models to be able to customize the component). And for controllers we could add a configuration to add a regex to extract the name of the models through the name of the controller. I believe that this way we could customize this extraction both the model directory and the controller suffix. Remembering that these points I went through were thinking about the convention adopted by Laravel, so unfortunately if the user chooses to create a structure that is very distant from the current structure, it will not be possible for us to use some tools. But if there is another option on how to get the model's path through the route/request/controller, I can make the modification

File: src/LaravelRequestDocs.php [hunk 10]
```diff
@@ -426,6 +460,7 @@ public function rulesByRegex(string $requestClassName, string $methodName): arra
                 return ['key' => $fieldName, 'rules' => $definedFieldRules];
             })
             ->keyBy('key')
+            // @phpstan-ignore-next-line
```
[the remark this answers is not in the document]
Comment (fernandopiovezan1) [reply src/LaravelRequestDocs.php:463]: This comment was added to skip phpstan, I didn't make changes to this part of the code, I just added the comment so that the pipeline wouldn't break like it was done a few lines above

File: config/request-docs.php (hunk 7 above)
Comment (fernandopiovezan1) [reply config/request-docs.php:185]: The definition can only be done through the format that the operator defines in the rule as you posted, but when winning as a team the rules are not always defined in the same order so we don't have a standard and at least in my opinion it looks messy in documents like In the print below, I added this option simply to organize the rules cohesively but there is still the possibility of not passing the ordering and thus returning as defined in the method
image

**Comment by fernandopiovezan1** [comment]
@kitloong I uploaded some changes so that the validations don't get too big, there are still a few points left that I should finish by the weekend. Any other notes I am at your disposal
Thanks

cc @kevincobain2000

File: tests/Stubs/TestControllers/WelcomeController.php (hunk 4 above)
Comment (fernandopiovezan1) [reply tests/Stubs/TestControllers/WelcomeController.php]: Foram criados testes para cobrir todas as novas implementações

File: src/LaravelRequestDocs.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (fernandopiovezan1) [reply src/LaravelRequestDocs.php:367]: I made an implementation where the pattern must be passed in the settings to extract the model name from the controller name and the factory directory must also be entered so the values ​​are not fixed and can be easily configured by the user

**Comment by fernandopiovezan1** [comment]
Hello @kevincobain2000 and @kitloong , I completed the requested adjustments, mostly as requested and others with some differences for correct operation. I will explain the reason for each of the points that left you with doubts.
field_info: for users who consume the API, a brief description of what the expected field is makes sense (although it is often intuitive). I analyzed several packages that offer this option and this was the most elegant alternative I found (there are packages that include a comment on top of each rule, which I found very strange, as we ended up with a dirty array) so the method, despite not being the best way to work, I can still use it in conjunction with the formRequest's attributes() method.
rules_order: this is a more ethical situation to maintain standardization independent of the developer. [more info](https://github.com/rakuntech/laravel-request-docs/pull/365#discussion_r1628345257)
examples: This is a point that makes a lot of sense to me, my frontend developers always ask about the real values ​​that should be sent to the API and we always need to manually inform these values ​​since we have factories that generate these values ​​and are ignored in the time to provide these examples. Using the way I implemented it is possible to customize both the way that "captures" the model name and the directory where the factories are. And there is still the possibility of disabling its use, without causing any impact on what is already implemented.
The tests were implemented and I believe it provided the necessary coverage and security for operation.

**Comment by kevincobain2000** [comment]
Thanks @fernando-piovezan-instar 
Please give us some time, we are on it checking.

**Comment by kitloong** [comment]
Thank you @fernandopiovezan1 

In `field_info` we have example.

```
    return [
            'name' => [
                'description' => 'Name of user',
                'example' => 'Jhon Doe'
            ],
            'password' => [
                'description' => 'The password of the user being created',
                'example' => '###123@@@890'
            ],
            'remember' => [
                'description' => 'Whether or not the user should be reminded'
            ],
        ];
```

Then, we have factory example in `example`, from your description, it is example value for model property, looks to me it is similar to the `example` in `field_info`.

How this 2 works together, and could you provide a screenshot from your side for different scenario?
eg:
1. User provide `field_info` example only
2. User provide factory example only
3. User provide both example at the same time.

**Comment by fernando-piovezan-instar** [comment]
Hi @kitloong There really were a lot of changes and I ended up forgetting to update the readme (I updated it now), the example key was removed from the fieldDescriptions() method, this method will only return the field description, the example will still be used but it will come directly from which was already generated by the factory, so it eliminates the duplication of examples, I believe that in this format the code becomes cleaner, the fieldDescription was very similar to the attributes() method of the formRequest (I considered using it but as it is a description validation messages look strange so I kept the description method). Sorry for not updating the readme sooner.

Thanks

**Comment by fernandopiovezan1** [comment]
Hi @kitloong There really were a lot of changes and I ended up forgetting to update the readme (I updated it now), the example key was removed from the fieldDescriptions() method, this method will only return the field description, the example will still be used but it will come directly from which was already generated by the factory, so it eliminates the duplication of examples, I believe that in this format the code becomes cleaner, the fieldDescription was very similar to the attributes() method of the formRequest (I considered using it but as it is a description validation messages look strange so I kept the description method). Sorry for not updating the readme sooner.
Below is a printout of the situations you raised

image

Thanks

**Comment by kitloong** [comment]
Hi @fernandopiovezan1 , sorry for the late reply and thank you for the screenshots.

First of all,

I think extracting the model from the controller name and factory name is not a good implementation to generate the example.
In most cases, your API request fields aren't always the same as your factory properties, hence I see it doesn't fit for general purpose.

However, I am convinced we could provide examples at: 

```
        return [
            'name' => [
                'description' => 'Name of user',
                'example' => 'Jhon Doe'
            ],
            'password' => [
                'description' => 'The password of the user being created',
                'example' => '###123@@@890'
            ],
            'remember' => [
                'description' => 'Whether or not the user should be reminded'
            ],
        ];
```

It makes more sense as it follows the request fields.

Second, LRD should display the real ordering defined in `rules` method. If you want to keep a consistent ordering, you should set and control your ordering in the `rules` method. Using `rules_order` to display a different ordering will create confusion, as it doesn't reflect the real rule ordering.

In summary, I am okay to add the following features:

1. `summary`
1. `description`
1. `field_info`
1. `tag`

**Comment by fernando-piovezan-instar** [comment]
Hi @kitloong, I understand from your point of view, despite not agreeing, mainly with the question of factory nomenclatures, that Laravel instructed us to use exactly the form that was implemented Controller, Models, Factories, etc. following the same nomenclature, changing only or suffix. So by convention you have to disagree with your opinion.
About the use of examples was also confusing, because the implementation of the examples described was not an option and now it? I also believe that for the most part the situations in the factories are better than the fact that the factories must be used to generate masses for tests and consequently these tests must be sent to the broken entry for validation of the rules and I do not understand why the factories We are not an alternative boa. If they do not exist, nothing will be generated either.
As for real order, it would simply be to leave the empty array that everything would be displayed as it was created (I don't particularly like it because I prefer everything properly standardized).
At the moment I am not able to make the requested alterations because I am doing a lot of work, if I want to date the PR so as not to be able to confirm it, in the future when the conditions are returned to send the PR to you. Thank you for your attention and for the time taken and once again thank you for the parabenization for the very best tool that we created and made available and in case you need the alterations mentioned or fork in my repository is working with the various alterations in the master branch https://github.com/fernandopiovezan1/laravel-request-docs/tree/master

cc @kevincobain2000 

Thanks

**Comment by kevincobain2000** [comment]
Thanks for all the efforts. 

Will cherry pick the commits that @kitloong mentioned. I also agree with not including the factory and rules ordering.

**Comment by kevincobain2000** [comment]
> Laravel instructed us to use exactly the form that was implemented Controller, Models, Factories, etc. following the same nomenclature, changing only or suffix. So by convention you have to disagree with your opinion.

Really? When did that become a must? And is it on PHPStan or other Linters that can catch it?

**Comment by kevincobain2000** [comment]
Closing it as patch on a fork. 
May pick up the examples from phptag title and description on horizon.

Thanks for submitting.
